Rudimentary Gamepad Recording and Loading Support for FlxReplay

agmass opened the pull request:
Adds support for recording and loading gamepad inputs inside FlxReplay-s, so far only support button and analog inputs. 
Analog Inputs are always recorded even when 0 which may cause a small jump in replay file size when having multiple gamepad inputs connected. There are a few areas where I touch FlxGamepad and FlxGamepadManager that could seem a bit sketchy, but i'm not aware of any other ways to get this to work. Tested on CPP and HTML5

agmass committed:
start some work

agmass committed:
will this fix it

agmass committed:
finished?

agmass committed:
refactor some comments

agmass committed:
only save inputs when using gameinput_api

Geokureli commented:
Could you share a little test state that would allow me to verify its accuracy? Perhaps we can add Gampad controls to the [Replay demo](https://haxeflixel.com/demos/Replay/)?

agmass commented:
> Could you share a little test state that would allow me to verify its accuracy? Perhaps we can add Gampad controls to the [Replay demo](https://haxeflixel.com/demos/Replay/)?

I've been testing this on my own game with a bit more bloat, but here's a small video showing the system:

https://github.com/user-attachments/assets/5e65f425-283a-4efd-82ff-f42129a083cb

The debug text on the right has a "Input" section showing the exact input from the gamepad/demo, as some are ignored

> Could we get this to behave like keyboard? where zero values are ignored?

This could be possible if the last frame is somehow given to FlxGamepad#record(), i'll try to make something now to get this working

Geokureli commented:
Also can you post an example of the string data?

agmass commented on flixel/input/gamepad/FlxGamepad.hx:
```diff
@@ -702,6 +710,15 @@ class FlxGamepad implements IFlxDestroyable
 
 	function getAxisValue(AxisID:Int):Float
 	{
+		if (recording)
+		{
+			if (AxisID < 0 || AxisID >= axis.length)
+			{
+				return 0;
+			}
+			return axis[AxisID];
+		}
+		
```
I'm editing the axis list upon playback (i'm not entirely sure if there's a better way to edit the axises of a gamepad, that's just what i was able to find) but if i don't specify the gamepad has to find it's axis from there, the gamepad will pull it directly from the input device. Along with this, the recording could double-up the axis flip as getAxisValue() flips it once during playback and could flip it again if called raw when recording.

agmass committed:
don't give 0-axis values

agmass commented:
> Also can you post an example of the string data?
![image](https://github.com/user-attachments/assets/43e82bc4-ac2f-40dc-9b9c-75bbe5003ce1)
after the commit trimming down the size by not including all axis, even when disabled

Geokureli commented:
FYI all the CI jobs are failing, so I haven't started reviewing this

Geokureli commented on flixel/system/replay/FlxReplay.hx:
```diff
@@ -165,6 +166,19 @@ class FlxReplay
 			continueFrame = false;
 		#end
 
+		#if FLX_GAMEINPUT_API
+		var gamepadRecords:Array<GamepadRecord> = new Array();
+		for (gamepad in FlxG.gamepads.getActiveGamepads())
+		{
+			var gamepadRecord:GamepadRecord = gamepad.record();
+			if (gamepadRecord != null)
+			{
+				gamepadRecords.push(gamepadRecord);
+				continueFrame = false;
+			}
+		}
+		#end
```
```suggestion
		#if FLX_GAMEPAD
		var gamepadRecords:Array<GamepadRecord> = new Array();
		for (gamepad in FlxG.gamepads.getActiveGamepads())
		{
			var gamepadRecord:GamepadRecord = gamepad.record();
			if (gamepadRecord != null)
			{
				gamepadRecords.push(gamepadRecord);
				continueFrame = false;
			}
		}
		#end
```

Geokureli commented on flixel/system/replay/FlxReplay.hx:
```diff
@@ -178,6 +192,9 @@ class FlxReplay
 		#if FLX_KEYBOARD
 		frameRecorded.keys = keysRecord;
 		#end
+		#if FLX_GAMEINPUT_API
```
```suggestion
		#if FLX_GAMEPAD
```

Geokureli commented on flixel/system/replay/FlxReplay.hx:
```diff
@@ -220,6 +239,28 @@ class FlxReplay
 			FlxG.mouse.playback(fr.mouse);
 		}
 		#end
+		#if FLX_GAMEINPUT_API
```
```suggestion
		#if FLX_GAMEPAD
```

Geokureli committed:
change FLX_GAMEINPUT_API to FLX_GAMEPAD

FLX_GAMEPAD is the opposite of FLX_NO_GAMEPAD, where FLX_GAMEINPUT_API is the opposite of  FLX_JOYSTICK_API which, I think, uses a different input system

agmass commented:
I'm going to close this PR because it's been open on my account for months and I don't work on games much with haxeflixel anymore, someone else can pick this up if they'd like

Geokureli commented:
Thanks for the contribution, sorry for the long wait, I'll try to pick this up soon
